### PR TITLE
CephFS: support omap data store in radosnamespace

### LINKED ISSUE #4599: Support storing cephfs omap in a different namespace in rados under same filesystem metadata pool
# Describe the feature you'd like to have #

Support storing cephfs omap in a different namespace in rados under the same filesystem metadata pool

# What is the value to the end user? (why is it a priority?) #

With this feature, even the metadata of the pvc/subvolumes will be stored under a different namespace, where security/isolation is the priority the user/customers can configure csi to use a different namespace.

# How will we know we have a good solution? (acceptance criteria) #

We can have a new field in the configmap where we have an entry for the subvolumegroup name.

### PR DESCRIPTION
**Pull request by iPraveenParihar**
# Describe what this PR does #

This commit adds the support for storing the CephFS omap data
in a namespace specified in the `ceph-csi-config` ConfigMap under
cephFS.radosNamespace field.

If the radosNamespace is not set, the default radosNamespace will be
used i.e, `csi`.

## Related issues ##

Closes: #4599 

## Future concerns ##

**Checklist:**

* [x] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#commit-messages).
* [x] Reviewed the developer guide on [Submitting a Pull Request](https://github.com/ceph/ceph-csi/blob/devel/docs/development-guide.md#development-workflow)
* [x] [Pending release notes](https://github.com/ceph/ceph-csi/blob/devel/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next major release.
* [x] Documentation has been updated, if necessary.
* [x] Unit tests have been added, if necessary.
* [x] Integration tests have been added, if necessary.

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

* `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)

</details>

### CODE REVIEW

**Comment by iPraveenParihar** [comment]
Initial test

>  configMap
```json
{"clusterID":"rn","monitors":["10.96.180.5:6789"],"cephFS":{"radosNamespace":"rn","netNamespaceFilePath":"","subvolumeGroup":"","kernelMountOptions":"","fuseMountOptions":""},"rbd":{"netNamespaceFilePath":"","radosNamespace":""},"nfs":{"netNamespaceFilePath":""},"readAffinity":{"enabled":false,"crushLocationLabels":
    []},"namespace":""}
```

> CephFS PVC created
```bash
pm@dhcp53-176:~$ k get pv pvc-983953db-919c-4f5d-a33e-9b59cdf80246 -o jsonpath='{.spec.csi.volumeAttributes.subvolumeName}'
csi-vol-c0ef71cc-8daa-47a4-953e-277095874736 
```

>  volume omap data in radosNamespace `rn`

```bash
[root@c1 /]# rados ls -p myfs-metadata --namespace rn
csi.volume.c0ef71cc-8daa-47a4-953e-277095874736
csi.volumes.default
[root@c1 /]# rados listomapvals csi.volume.c0ef71cc-8daa-47a4-953e-277095874736 -p myfs-metadata --namespace rn
csi.imagename
value (44 bytes) :
00000000  63 73 69 2d 76 6f 6c 2d  63 30 65 66 37 31 63 63  |csi-vol-c0ef71cc|
00000010  2d 38 64 61 61 2d 34 37  61 34 2d 39 35 33 65 2d  |-8daa-47a4-953e-|
00000020  32 37 37 30 39 35 38 37  34 37 33 36              |277095874736|
0000002c

csi.volname
value (40 bytes) :
00000000  70 76 63 2d 39 38 33 39  35 33 64 62 2d 39 31 39  |pvc-983953db-919|
00000010  63 2d 34 66 35 64 2d 61  33 33 65 2d 39 62 35 39  |c-4f5d-a33e-9b59|
00000020  63 64 66 38 30 32 34 36                           |cdf80246|
00000028

csi.volume.owner
value (12 bytes) :
00000000  63 65 70 68 66 73 2d 72  61 64 6f 73              |cephfs-rados|
0000000c

[root@c1 /]# rados listomapvals csi.volumes.default -p myfs-metadata --namespace rn
csi.volume.pvc-983953db-919c-4f5d-a33e-9b59cdf80246
value (36 bytes) :
00000000  63 30 65 66 37 31 63 63  2d 38 64 61 61 2d 34 37  |c0ef71cc-8daa-47|
00000010  61 34 2d 39 35 33 65 2d  32 37 37 30 39 35 38 37  |a4-953e-27709587|
00000020  34 37 33 36                                       |4736|
00000024
[root@c1 /]#
```

> CephFS snapshot created
```bash
pm@dhcp53-176:~$ k get vs cephfs-pvc-snap-rn -o jsonpath='{.status.boundVolumeSnapshotContentName}'
snapcontent-0b4c4472-b33b-4c85-af43-76cdd0c34375 
```

>  CephFS snapshot omap data
```bash
[root@c1 /]# rados ls -p myfs-metadata --namespace rn
csi.snap.510ab77e-fbe9-4a51-ad0d-0644c4616678
csi.snaps.default
csi.volume.c0ef71cc-8daa-47a4-953e-277095874736
csi.volumes.default
[root@c1 /]# rados listomapvals csi.snap.510ab77e-fbe9-4a51-ad0d-0644c4616678 -p myfs-metadata --namespace rn
csi.imagename
value (45 bytes) :
00000000  63 73 69 2d 73 6e 61 70  2d 35 31 30 61 62 37 37  |csi-snap-510ab77|
00000010  65 2d 66 62 65 39 2d 34  61 35 31 2d 61 64 30 64  |e-fbe9-4a51-ad0d|
00000020  2d 30 36 34 34 63 34 36  31 36 36 37 38           |-0644c4616678|
0000002d

csi.snapname
value (45 bytes) :
00000000  73 6e 61 70 73 68 6f 74  2d 30 62 34 63 34 34 37  |snapshot-0b4c447|
00000010  32 2d 62 33 33 62 2d 34  63 38 35 2d 61 66 34 33  |2-b33b-4c85-af43|
00000020  2d 37 36 63 64 64 30 63  33 34 33 37 35           |-76cdd0c34375|
0000002d

csi.source
value (44 bytes) :
00000000  63 73 69 2d 76 6f 6c 2d  63 30 65 66 37 31 63 63  |csi-vol-c0ef71cc|
00000010  2d 38 64 61 61 2d 34 37  61 34 2d 39 35 33 65 2d  |-8daa-47a4-953e-|
00000020  32 37 37 30 39 35 38 37  34 37 33 36              |277095874736|
0000002c

[root@c1 /]# rados listomapvals csi.snaps.default -p myfs-metadata --namespace rn
csi.snap.snapshot-0b4c4472-b33b-4c85-af43-76cdd0c34375
value (36 bytes) :
00000000  35 31 30 61 62 37 37 65  2d 66 62 65 39 2d 34 61  |510ab77e-fbe9-4a|
00000010  35 31 2d 61 64 30 64 2d  30 36 34 34 63 34 36 31  |51-ad0d-0644c461|
00000020  36 36 37 38                                       |6678|
00000024
[root@c1 /]#
```

File: api/deploy/kubernetes/csi-config-map.go [hunk 1]
```diff
@@ -36,6 +36,8 @@ type CephFS struct {
 	NetNamespaceFilePath string `json:"netNamespaceFilePath"`
 	// SubvolumeGroup contains the name of the SubvolumeGroup for CSI volumes
 	SubvolumeGroup string `json:"subvolumeGroup"`
+	// RadosNamespace is a rados namespace in the pool
```
Comment (Madhu-1) [inline api/deploy/kubernetes/csi-config-map.go]: in the filesystem metadata pool

**Review by Madhu-1 — COMMENTED** [review]
please update the below as well.

* Helm charts for this one.
* Documentation (not to change this for existing clusterID)
* Release notes.

**Comment by Madhu-1** [comment]
@iPraveenParihar can you please rebase this PR

File: e2e/cephfs.go [hunk 2]
```diff
@@ -427,6 +427,90 @@ var _ = Describe(cephfsType, func() {
 				}
 			})
 
+			By("verify rados objects are within a namespace", func() {
+				framework.Logf("updating configmap with rados namespace %s", radosNamespace)
+				radosNamespace = "cephfs-ns"
```
Comment (nixpanic) [inline e2e/cephfs.go]: Poeh, this is a global variable, changing it during runtime can possibly have unexpected side-effects. Can you add a comment that it is critical that the value is restored to whatever it was (maybe not `""`?) at the end of this test. 

Or, combine the changing with restoring together, like so:

```go
oldRadosNamespace := radosNamespace
radosNamespace = "cephfs-ns"
defer func() {
    radosNamespace = oldRadosNamespace
}
```

**Comment by nixpanic** [comment]
Still need to address [this comment](https://github.com/ceph/ceph-csi/pull/4661#pullrequestreview-2235267139).

**Comment by iPraveenParihar** [comment]
> please update the below as well.
> 
> * Helm charts for this one.
> * Documentation (not to change this for existing clusterID)
> * Release notes.

@Madhu-1, This PR adds the field radosNamespace for CephFS struct in `ceph-csi-config` CM. There is no changes required Helm charts.
https://github.com/ceph/ceph-csi/blob/88b964fe18d37af5b53cea00e8d785159fb61f59/charts/ceph-csi-cephfs/values.yaml#L20-L31
Do you want me to add in example section? Or did I get it wrong?

**Comment by iPraveenParihar** [comment]
passed e2e [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.31/detail/mini-e2e_k8s-1.31/115/pipeline)

File: deploy/csi-config-map-sample.yaml [hunk 3]
```diff
@@ -79,6 +84,7 @@ data:
           "netNamespaceFilePath": "<kubeletRootPath>/plugins/cephfs.csi.ceph.com/net",
           "kernelMountOptions": "<kernelMountOptions for cephFS volumes>",
           "fuseMountOptions": "<fuseMountOptions for cephFS volumes>"
```
Comment (Madhu-1) [inline deploy/csi-config-map-sample.yaml]: add a comma to above fuse mountOptions at the end or else json parsing will have problem

File: e2e/cephfs.go [hunk 4]
```diff
@@ -2491,6 +2491,83 @@ var _ = Describe(cephfsType, func() {
...
+				err = deleteResource(cephFSExamplePath + "storageclass.yaml")
+				if err != nil {
+					framework.Failf("failed to delete CephFS storageclass: %v", err)
+				}
+
+				validateOmapCount(f, 0, cephfsType, metadataPool, volumesType)
+				err = deleteResource(cephFSExamplePath + "storageclass.yaml")
+				if err != nil {
+					framework.Failf("failed to delete CephFS storageclass: %v", err)
+				}
```
Comment (Madhu-1) [inline e2e/cephfs.go:2603]: This is not required as storageclass is already deleted above

File: e2e/cephfs.go [hunk 5]
```diff
@@ -2491,6 +2491,83 @@ var _ = Describe(cephfsType, func() {
...
+				// create a PVC and bind it to an app
+				pvc, pod, err := createPVCAndAppBinding(pvcPath, appPath, f, deployTimeout)
+				if err != nil {
+					framework.Failf("failed to validate CephFS pvc and application binding: %v", err)
+				}
+
+				validateOmapCount(f, 1, cephfsType, metadataPool, volumesType)
+
+				// delete resources
+				err = deletePod(pod.Name, pod.Namespace, f.ClientSet, deployTimeout)
```
Comment (Madhu-1) [inline e2e/cephfs.go:2584]: please add a test for snapshot and Restore pvc as well as we will use this new radosnamespace for metadata as well.

File: internal/util/csiconfig.go [hunk 6]
```diff
@@ -106,6 +107,21 @@ func GetRadosNamespace(pathToConfig, clusterID string) (string, error) {
...
+// GetCephFSRadosNamespace returns the namespace for the given clusterID.
+// If not set, it returns the default value "csi".
+func GetCephFSRadosNamespace(pathToConfig, clusterID string) (string, error) {
+	cluster, err := readClusterInfo(pathToConfig, clusterID)
+	if err != nil {
+		return "", err
+	}
+
+	if cluster.CephFS.RadosNamespace == "" {
+		return fsutil.RadosNamespace, nil
```
Comment (Madhu-1) [inline internal/util/csiconfig.go]: try to avoid importing fsutil in this standard util as it can create a circular dependency problem in the future.

**Review by Madhu-1 — COMMENTED** [review]
LGTM, some nits

**Comment by Madhu-1** [comment]
> the field radosNamespace for CephFS struct in

Yes please add a example in struct in helm charts as well

**Comment by iPraveenParihar** [comment]
passed e2e [logs](https://jenkins-ceph-csi.apps.ocp.cloud.ci.centos.org/job/mini-e2e_k8s-1.31/130/display/redirect)

File: charts/ceph-csi-cephfs/values.yaml [hunk 7]
```diff
@@ -28,6 +28,7 @@ serviceAccounts:
 #     cephFS:
 #       subvolumeGroup: "csi"
 #       netNamespaceFilePath: "{{ .kubeletDir }}/plugins/{{ .driverName }}/net"
+#       radosNamespace: "app-1"
```
Comment (Madhu-1) [inline charts/ceph-csi-cephfs/values.yaml]: `app-1` to `csi` lets use some meaningful names

**Review by Madhu-1 — COMMENTED** [review]
small nit, LGTM

File: charts/ceph-csi-cephfs/values.yaml (hunk 7 above)
Comment (iPraveenParihar) [reply charts/ceph-csi-cephfs/values.yaml]: done

**Review by Madhu-1 — DISMISSED** [review]

**Comment by iPraveenParihar** [comment]
@nixpanic @Rakshith-R PTAL

**Review by Rakshith-R — DISMISSED** [review]
LGTM

**Comment by Rakshith-R** [comment]
@iPraveenParihar Can you please rebase and resolve conflicts ?

**Review by Rakshith-R — APPROVED** [review]

**Review by Madhu-1 — APPROVED** [review]

**Comment by Rakshith-R** [comment]
@mergifyio queue

**Pull request merged**